[stdlib] Faster repr() logic for String, StringRef, UInt, DType and others

msaelices opened the pull request:
- [X] Implement `StringRef.__repr__()`
- [X] Faster `path.stat()` implementation using Writer
- [X] Faster `UInt.__repr__()`
- [X] Faster `Error.__repr__()`
- [X] Faster `DType.__repr__()`

msaelices committed:
Implement StringRef.__repr__()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Faster path.stat() implementation using Writer

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Optimize String.__repr__() by using Writer

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Use StringRepr for StringRef.__repr__()
This improves the representation to be like String.__repr__()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Add changelog entry

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

JoeLoser reviewed (commented):
Do you mind looking into the pathlib test failure as a result of this change?  See [this CI run](https://github.com/modularml/mojo/actions/runs/11655733986/job/32497162367?pr=3736) for example.

msaelices committed:
[stdlib] Faster UInt.__repr__()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Faster Error.__repr__()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Faster DType.__repr__()

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Simpler stat_result.__repr__() logic

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Hopefully fix the test_stat() test in MacOS

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into better-representable

msaelices committed:
[stdlib] Simpler String.__repr__() logic

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Remove two failing checks in test_stat() which depends on the OS

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Merge branch 'nightly' into better-representable

martinvuyk reviewed (changes requested):
Hi, I've been working on faster repr and ascii logic for about 3 weeks now, see PR #3686. This does not add much speed really. 

I think you can still merge the other improvements you made on this PR. But see the [_repr function](https://github.com/modularml/mojo/pull/3686/files#diff-53bade406e0c0f5616a9bbee7766e2f85b9ab8cbf1acd7441d3e52aef555465cR261) that I've been working on, there are several other things that need to exist as helpers to make this much faster.

msaelices committed:
Merge branch 'nightly' into better-representable

msaelices committed:
[stdlib] Adapt the code to s/unsafe_from_utf8_ptr/ptr refactoring

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Replace StringRepr to _StringRepr

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
Fix typo in _StringRepr docstring

Co-authored-by: Joe Loser <joeloser@fastmail.com>
Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Do not use _StringRepr for StringRef.__repr__()
We will take advantage of a faster StringRef.__repr__() when
https://github.com/modularml/mojo/pull/3686 was merged

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

msaelices committed:
[stdlib] Revert the StringRepr logic as there is a much better approach
waiting for being merged here: https://github.com/modularml/mojo/pull/3686

Signed-off-by: Manuel Saelices <msaelices@gmail.com>

martinvuyk reviewed (approved):
LGTM
